### PR TITLE
RSDK-11613 —       Update Go camera API with new GetImages signature

### PR DESCRIPTION
**Pull request by hexbabe**
[RSDK-11613](https://viam.atlassian.net/browse/RSDK-11613)

Todo:
- [x] Write better unit tests instead of just getting stuff to compile
- [x] Remove POC stream server changes (deferred to [another ticket](https://viam.atlassian.net/browse/RSDK-11641))
- [x] Manual testing

# Manual Tests

## Data Manager GetImages 
Re-ran this test https://github.com/viamrobotics/rdk/pull/5209#issuecomment-3188903120 exactly as outlined and made sure it still works off of this feature branch.

## DataManager ReadImage
Changed config to use `"method": "ReadImage"` instead of `"method": "GetImages"`, filtered_camera still works as expected.

## Stream server
Live streams working as expected.

## GetImages filter_source_names
Using a go client script to call `Images` on an `image_file` camera with below config
```
    {
      "name": "image-file-1",
      "api": "rdk:component:camera",
      "model": "rdk:builtin:image_file",
      "attributes": {
        "preloaded_image": "pizza",
        "color_image_file_path": "/Users/seanyu/Downloads/image.png"
      }
    }
```

### Requesting invalid source 
```
2025-08-19T18:53:05.415Z        ERROR   client  go-sdk-client/client.go:50      camera client: could not gets images from the camera rpc error: code = Unknown desc = camera server GetImages could not call Images on the camera: invalid source name: RAWWWWWWRRRRRR
```

### Requesting with no sources specified
```
	for _, image := range namedImages {
		logger.Info(image.MimeType())
	}
	logger.Info(respMetadata)
```
Logged: 
```
2025-08-19T18:54:38.159Z        INFO    client  go-sdk-client/client.go:53      image/jpeg
2025-08-19T18:54:38.159Z        INFO    client  go-sdk-client/client.go:53      image/jpeg
2025-08-19T18:54:38.159Z        INFO    client  go-sdk-client/client.go:55      {2025-08-19 18:54:38.149384 +0000 UTC}
```

### Requesting source "preloaded" specified
```
2025-08-19T18:56:30.575Z        INFO    client  go-sdk-client/client.go:53      image/jpeg
2025-08-19T18:56:30.575Z        INFO    client  go-sdk-client/client.go:55      {2025-08-19 18:56:30.572539 +0000 UTC}
```
Number of sources checks out.

### Requesting duplicate sources "color" and "color"
```
2025-08-19T18:57:30.835Z        ERROR   client  go-sdk-client/client.go:50      duplicate source name in filter: color
exit status 1
```
Nice.

[RSDK-11613]: https://viam.atlassian.net/browse/RSDK-11613?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: rimage/image_file.go [hunk 1]
```diff
@@ -47,6 +47,9 @@ const RawDepthHeaderLength = 24
 func init() {
 	// Here we register the custom format above so that we can simply use image.Decode
 	// so long as the raw RGBA data has the appropriate header
+	// Make sure that format registration names e.g. "vnd.viam.rgba" are trailing substrings of
+	// its corresponding MIME type e.g. "image/vnd.viam.rgba" in mime.go. This is crucial to make sure
+	// that our mime type handling is 1:1 with the registered formats.
```
Comment (hexbabe) [inline rimage/image_file.go:52]: Result of discussion and investigation [here](https://viaminc.slack.com/archives/C01TBQ72NBX/p1755199021819309)

File: components/camera/camera.go [hunk 2]
```diff
@@ -75,8 +77,78 @@ type Properties struct {
...
+	}
+	if mimeType == "" {
+		return NamedImage{}, fmt.Errorf("must provide a mime type to construct a named image")
+	}
+	return NamedImage{img: img, SourceName: sourceName, mimeType: mimeType}, nil
+}
+
+// Image returns the image.Image of the NamedImage.
+func (ni *NamedImage) Image(ctx context.Context) (image.Image, error) {
+	if ni.img == nil {
```
Comment (seanavery) [inline components/camera/camera.go]: I think it could be clearer if you immediately returned `ni.img` if it is not nil at the beginning here

File: components/camera/camera.go [hunk 3]
```diff
@@ -75,8 +77,78 @@ type Properties struct {
...
+		}
+
+		reader := bytes.NewReader(ni.data)
+		_, header, err := image.DecodeConfig(reader)
+		if err != nil {
+			return nil, fmt.Errorf("could not decode image config: %w", err)
+		}
+
+		if header != "" && !strings.Contains(ni.mimeType, header) {
+			return nil, fmt.Errorf("mime type does not match the image bytes: expected %s, got %s", ni.mimeType, header)
```
Comment (seanavery) [inline components/camera/camera.go]: Why would we not want to update the mimeType to ground truth and return success in the case they do not match?

Sorry if this was already hashed out in the scope doc.

File: components/camera/client.go [hunk 4]
```diff
@@ -227,23 +231,14 @@ func (c *client) Images(ctx context.Context, extra map[string]interface{}) ([]Na
...
-			rdkImage = rimage.NewLazyEncodedImage(img.Image, utils.MimeTypeJPEG)
-		case pb.Format_FORMAT_PNG:
-			rdkImage = rimage.NewLazyEncodedImage(img.Image, utils.MimeTypePNG)
-		case pb.Format_FORMAT_UNSPECIFIED:
-			rdkImage, _, err = image.Decode(bytes.NewReader(img.Image))
-			if err != nil {
-				return nil, resource.ResponseMetadata{}, err
-			}
+		if img.MimeType == "" {
+			img.MimeType = utils.FormatToMimeType[img.Format]
```
Comment (seanavery) [inline components/camera/client.go:237]: Is it possible for the mimetype lookup here to fail?

**Review by seanavery — COMMENTED** [review]
First pass, only looked at the components/camera diff

File: components/camera/camera.go (hunk 3 above)
Comment (hexbabe) [reply components/camera/camera.go]: This was hashed out in an offline convo with me, Nick, and Bijan so no worries

Basically we want to tell the server about its bug i.e. the mime type not matching the bytes. It's a tradeoff because as you said it could prevent the downstream caller to receive perfectly healthy bytes. But yeah it's an intentional design decision.

Originally I proposed it to make the ground truth conversion and warn log, but that would require passing a logger into `Image` which adds extra overhead and usability difficulties.

Edit: Honestly after thinking about it though, we could return a specific error type, which we can then catch in the server/client

File: components/camera/client.go (hunk 4 above)
Comment (hexbabe) [reply components/camera/client.go:237]: I don't think so. Format is a int32 typed enum, so the default falsy value is `	Format_FORMAT_UNSPECIFIED Format = 0`. And the `utils.FormatToMimeType` map covers all enum cases.

File: components/camera/camera.go (hunk 2 above)
Comment (hexbabe) [reply components/camera/camera.go]: Addressed here https://github.com/viamrobotics/rdk/pull/5183/commits/a90370f8989a55438f8a3488f95079bf42cd8677

File: components/camera/client.go (hunk 4 above)
Comment (hexbabe) [reply components/camera/client.go:237]: Action: add TODO comment that this will be removed soon and add context on why we are adding it

File: components/camera/camera.go [hunk 5]
```diff
@@ -179,30 +264,38 @@ func DecodeImageFromCamera(ctx context.Context, mimeType string, extra map[strin
 func GetImageFromGetImages(
```
Comment (hexbabe) [inline components/camera/camera.go:266]: update above comment to say this will be deprecated soon and why it was needed

File: components/camera/camera.go [hunk 6]
```diff
@@ -249,12 +342,12 @@ func GetImagesFromGetImage(
 		logger.Warnf("requested mime type %s, but received %s", mimeType, resMimetype)
 	}
 
-	img, err := rimage.DecodeImage(ctx, resBytes, utils.WithLazyMIMEType(resMetadata.MimeType))
```
Comment (hexbabe) [inline components/camera/camera.go:252]: add comment to func docstring that this method is deprecated and why it was needed

File: components/camera/camera_test.go [hunk 7]
```diff
@@ -287,94 +291,113 @@ func TestGetImageFromGetImages(t *testing.T) {
...
+		extra map[string]interface{},
+	) ([]camera.NamedImage, resource.ResponseMetadata, error) {
+		namedImg, err := camera.NamedImageFromImage(dm, source1Name, rutils.MimeTypeRawDepth)
+		if err != nil {
+			return nil, resource.ResponseMetadata{}, err
+		}
+		return []camera.NamedImage{namedImg}, resource.ResponseMetadata{CapturedAt: time.Now()}, nil
 	}
 
 	t.Run("PNG mime type", func(t *testing.T) {
```
Comment (hexbabe) [inline components/camera/camera_test.go]: please remove thsi test

File: components/camera/camera_test.go [hunk 8]
```diff
@@ -287,94 +291,113 @@ func TestGetImageFromGetImages(t *testing.T) {
...
-		imgBytes, metadata, err := camera.GetImageFromGetImages(context.Background(), nil, rutils.MimeTypePNG, rgbaCam, nil)
+		imgBytes, metadata, err := camera.GetImageFromGetImages(context.Background(), nil, rgbaCam, nil, nil)
 		test.That(t, err, test.ShouldBeNil)
-		test.That(t, metadata.MimeType, test.ShouldEqual, rutils.MimeTypePNG)
+		test.That(t, metadata.MimeType, test.ShouldEqual, rutils.MimeTypeRawRGBA)
 		verifyDecodedImage(t, imgBytes, rutils.MimeTypePNG, testImg1)
 	})
 
 	t.Run("JPEG mime type", func(t *testing.T) {
```
Comment (hexbabe) [inline components/camera/camera_test.go]: remove this test

File: components/camera/camera_test.go [hunk 9]
```diff
@@ -460,31 +489,317 @@ func TestGetImagesFromGetImage(t *testing.T) {
 	})
 }
 
-func TestImagesExtraParam(t *testing.T) {
+func TestImages(t *testing.T) {
```
Comment (hexbabe) [inline components/camera/camera_test.go:471]: add comment about the responsibilites of this test

File: components/camera/camera_test.go [hunk 10]
```diff
@@ -460,31 +489,317 @@ func TestGetImagesFromGetImage(t *testing.T) {
...
+				if img, ok := availableSources[sourceName]; ok {
+					result = append(result, img)
+				} else {
+					return nil, resource.ResponseMetadata{}, fmt.Errorf("requested source name not found: %s", sourceName)
+				}
+			}
+			return result, resource.ResponseMetadata{}, nil
+		}
+
+		t.Run("no filter returns all sources", func(t *testing.T) {
```
Comment (hexbabe) [inline components/camera/camera_test.go]: change to "nil filter"

File: components/camera/camera_test.go [hunk 11]
```diff
@@ -460,31 +489,317 @@ func TestGetImagesFromGetImage(t *testing.T) {
...
+			test.That(t, len(imgs), test.ShouldEqual, 3)
+
+			returnedSources := make(map[string]bool)
+			for _, img := range imgs {
+				returnedSources[img.SourceName] = true
+			}
+			test.That(t, returnedSources[source1Name], test.ShouldBeTrue)
+			test.That(t, returnedSources[source2Name], test.ShouldBeTrue)
+			test.That(t, returnedSources[source3Name], test.ShouldBeTrue)
+			test.That(t, len(returnedSources), test.ShouldEqual, 3)
```
Comment (hexbabe) [inline components/camera/camera_test.go]: move this before the truthy asserts

File: components/camera/camera_test.go [hunk 12]
```diff
@@ -460,31 +489,317 @@ func TestGetImagesFromGetImage(t *testing.T) {
...
```
Comment (hexbabe) [inline components/camera/camera_test.go]: move this before the truthy asserts

File: components/camera/camera_test.go [hunk 13]
```diff
@@ -460,31 +489,317 @@ func TestGetImagesFromGetImage(t *testing.T) {
...
+	testImgPNGBytes, err := rimage.EncodeImage(ctx, testImg, rutils.MimeTypePNG)
+	test.That(t, err, test.ShouldBeNil)
+	testImgJPEGBytes, err := rimage.EncodeImage(ctx, testImg, rutils.MimeTypeJPEG)
+	test.That(t, err, test.ShouldBeNil)
+	badBytes := []byte("trust bro i'm an image ong")
+	sourceName := "test_source"
+
+	t.Run("NamedImageFromBytes", func(t *testing.T) { //nolint:dupl
+		t.Run("success", func(t *testing.T) {
+			ni, err := camera.NamedImageFromBytes(testImgPNGBytes, sourceName, rutils.MimeTypePNG)
```
Comment (hexbabe) [inline components/camera/camera_test.go:638]: assert on image equality

File: components/camera/camera_test.go [hunk 14]
```diff
@@ -460,31 +489,317 @@ func TestGetImagesFromGetImage(t *testing.T) {
...
+		})
+		t.Run("error on empty mime type", func(t *testing.T) {
+			_, err := camera.NamedImageFromBytes(testImgPNGBytes, sourceName, "")
+			test.That(t, err, test.ShouldBeError, errors.New("must provide a mime type to construct a named image"))
+		})
+	})
+
+	t.Run("NamedImageFromImage", func(t *testing.T) { //nolint:dupl
+		t.Run("success", func(t *testing.T) {
+			ni, err := camera.NamedImageFromImage(testImg, sourceName, rutils.MimeTypePNG)
```
Comment (hexbabe) [inline components/camera/camera_test.go:655]: assert on image equality

File: components/camera/client_test.go [hunk 15]
```diff
@@ -366,12 +386,105 @@ func TestClient(t *testing.T) {
...
+
+		t.Run("successful filtering", func(t *testing.T) {
+			images, _, err := camClient.Images(ctx, []string{"color", "depth"}, nil)
+			test.That(t, err, test.ShouldBeNil)
+			test.That(t, len(images), test.ShouldEqual, 2)
+			test.That(t, images[0].SourceName, test.ShouldEqual, "color")
+			test.That(t, images[1].SourceName, test.ShouldEqual, "depth")
+
+			// Verify the actual image content
+			colorImg, err := images[0].Image(ctx)
```
Comment (hexbabe) [inline components/camera/client_test.go:449]: be thorough and evaluate exact pixel equality similar to camera_test.go

also add to the top of the file that these are integration tests that actually spin up a local socket connection

File: components/camera/client_test.go [hunk 16]
```diff
@@ -366,12 +386,105 @@ func TestClient(t *testing.T) {
...
+			images, _, err = camClient.Images(ctx, nil, nil)
+			test.That(t, err, test.ShouldBeNil)
+			test.That(t, len(images), test.ShouldEqual, 2)
+			colorImg, err = images[0].Image(ctx)
+			test.That(t, err, test.ShouldBeNil)
+			test.That(t, colorImg.Bounds().Dx(), test.ShouldEqual, 40)
+			test.That(t, colorImg.Bounds().Dy(), test.ShouldEqual, 50)
+			depthImg, err = images[1].Image(ctx)
+			test.That(t, err, test.ShouldBeNil)
+			test.That(t, depthImg.Bounds().Dx(), test.ShouldEqual, 10)
```
Comment (hexbabe) [inline components/camera/client_test.go:489]: same thing for this test, be thorough and evaluate exact pixel equality similar to camera_test.go

File: components/camera/camera.go (hunk 3 above)
Comment (seanavery) [reply components/camera/camera.go]: [ErrMIMETypeBytesMismatch](https://github.com/viamrobotics/rdk/pull/5183/files#diff-4eab9b93978c09650defc23c4274a4c77d87cfdea7775d86dcf8c838ec207a01R134) diff looks good

File: components/camera/collectors.go [hunk 17]
```diff
@@ -157,14 +157,14 @@ func newGetImagesCollector(resource interface{}, params data.CollectorParams) (d
...
-			format, imgBytes, err := encodeImageFromUnderlyingType(ctx, img.Image)
+			imgBytes, err := img.Bytes(ctx)
 			if err != nil {
 				return res, err
 			}
 			binaries = append(binaries, data.Binary{
 				Annotations: data.Annotations{Classifications: []data.Classification{{Label: img.SourceName}}},
 				Payload:     imgBytes,
-				MimeType:    data.CameraFormatToMimeType(format),
+				MimeType:    data.MimeTypeStringToMimeType(img.MimeType()),
```
Comment (seanavery) [inline components/camera/collectors.go:167]: [MimeTypeStringToMimeType](https://github.com/viamrobotics/rdk/pull/5183/files#diff-67c0823aa24d3667ac603603cf934d7c0d988547e19a88cd2d48b3aa1337dc99R300) string->enum

Why would we want to store enum instead of string value here?
I thought we wanted to move in that direction of string val instead of the limited enum list

Do you know where we define the collector "API" for Classifications here?

Comment (hexbabe) [reply components/camera/collectors.go:167]: We have to because the data API still uses enums. There should be a follow up scope and project on data side.

Comment (hexbabe) [reply components/camera/collectors.go:167]: https://viam.atlassian.net/browse/RSDK-11761

Comment (hexbabe) [reply components/camera/collectors.go:167]: > Do you know where we define the collector "API" for Classifications here?

https://github.com/viamrobotics/api/blob/d89fa180143261c21a4453ee63c672f6319ccf01/proto/viam/app/datasync/v1/data_sync.proto#L72

^ data's image mime types as per convo with Devin H

Comment (seanavery) [reply components/camera/collectors.go:167]: Got it, thanks for the link and ticket

File: components/camera/client_test.go (hunk 15 above)
Comment (seanavery) [reply components/camera/client_test.go:449]: Why are these pixel comparisons important again? Is this to make sure we are not re-encoding with JPEG?

Comment (hexbabe) [reply components/camera/client_test.go:449]: I think that and also just to make sure that the images are the exact image we expect to be extra sure the test is working.

**Review by SebastianMunozP — APPROVED** [review]
Reviewed these changes synchronously, LGTM

File: components/camera/camera.go [hunk 18]
```diff
@@ -75,8 +87,81 @@ type Properties struct {
...
+	if ni.data != nil {
+		return ni.data, nil
+	}
+	if ni.img == nil {
+		return nil, fmt.Errorf("no image or image bytes available")
+	}
+
+	data, err := rimage.EncodeImage(ctx, ni.img, ni.mimeType)
+	if err != nil {
+		return nil, fmt.Errorf("could not encode image: %w", err)
```
Comment (bhaney) [inline components/camera/camera.go]: could you make this more descriptive by including what encoding was attempted? e.g. 
`"could not encode image with encoding %s : %w", ni.mimeType, err) `

File: components/camera/client_test.go [hunk 19]
```diff
@@ -366,12 +388,109 @@ func TestClient(t *testing.T) {
...
+			return images, resource.ResponseMetadata{}, nil
+		}
+
+		t.Run("successful filtering", func(t *testing.T) {
+			images, _, err := camClient.Images(ctx, []string{"color", "depth"}, nil)
+			test.That(t, err, test.ShouldBeNil)
+			test.That(t, len(images), test.ShouldEqual, 2)
+			test.That(t, images[0].SourceName, test.ShouldEqual, "color")
+			test.That(t, images[1].SourceName, test.ShouldEqual, "depth")
+
```
Comment (bhaney) [inline components/camera/client_test.go:447]: Question: if the filter was in the opposite order `[]string{"depth", "color"}`, would it be expected that `images[0]` is the depth image, and `images[1]` is the color image? Is the order of request required in the order of response? Can you make a test?

File: components/camera/client_test.go [hunk 20]
```diff
@@ -366,12 +388,109 @@ func TestClient(t *testing.T) {
...
+			test.That(t, rimage.ImagesExactlyEqual(colorImg, expectedColor), test.ShouldBeTrue)
+
+			depthImg, err := images[1].Image(ctx)
+			test.That(t, err, test.ShouldBeNil)
+			test.That(t, depthImg.Bounds().Dx(), test.ShouldEqual, 10)
+			test.That(t, depthImg.Bounds().Dy(), test.ShouldEqual, 20)
+			test.That(t, rimage.ImagesExactlyEqual(depthImg, expectedDepth), test.ShouldBeTrue)
+		})
+
+		t.Run("empty and nil filters", func(t *testing.T) {
```
Comment (bhaney) [inline components/camera/client_test.go:462]: can you add a few more more tests, which test a filter that doesn't exist, and duplicates? 

case 1: `[]string{"no_such_source"}`
case 2: `[]string{"color", "no_such_source"}`
case 3: `[]string{"color", "color"}`

File: components/camera/server.go [hunk 21]
```diff
@@ -134,46 +145,6 @@ func (s *serviceServer) GetImages(
...
-			format = pb.Format_FORMAT_RAW_DEPTH
-		case utils.MimeTypeRawRGBA:
-			format = pb.Format_FORMAT_RAW_RGBA
-		case utils.MimeTypeJPEG:
-			format = pb.Format_FORMAT_JPEG
-		case utils.MimeTypePNG:
-			format = pb.Format_FORMAT_PNG
-		default:
-		}
-		return format, v.RawData(), nil
```
Comment (bhaney) [inline components/camera/server.go:152]: Feels really nice getting to delete this

File: components/camera/camera.go (hunk 18 above)
Comment (hexbabe) [reply components/camera/camera.go]: great call I'll add

Comment (hexbabe) [reply components/camera/camera.go]: added

File: components/camera/client_test.go (hunk 19 above)
Comment (hexbabe) [reply components/camera/client_test.go:447]: How to handle order would be entirely on the module implementer, client and server should have no opinions.

File: components/camera/client_test.go (hunk 20 above)
Comment (hexbabe) [reply components/camera/client_test.go:462]: I have case 1 and 2 in camera_test.go

case 3 is not handled by the client, it is handled by the server

Comment (hexbabe) [reply components/camera/client_test.go:462]: I'll add a test for case 3 to make it more explicit though

Comment (hexbabe) [reply components/camera/client_test.go:462]: Added all three to be extra safe

File: components/camera/client_test.go (hunk 19 above)
Comment (hexbabe) [reply components/camera/client_test.go:447]: I'll add a test to make that clear
